From 267c46e4351ea6d0d9415258cb7daa3b666a71ec Mon Sep 17 00:00:00 2001
From: Connor Abbott <cwabbott0@gmail.com>
Date: Thu, 21 May 2026 11:59:26 -0400
Subject: [PATCH] tu: Keep track of PARTIALLY_BOUND_BIT in descriptor sets

This will be used to see if loads can be speculatable.

Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/42159>
(cherry picked from mesa main commit 267c46e4351ea6d0d9415258cb7daa3b666a71ec)
---
 src/freedreno/vulkan/tu_descriptor_set.cc | 12 ++++++++++++
 src/freedreno/vulkan/tu_descriptor_set.h  |  6 ++++++
 2 files changed, 18 insertions(+)

diff --git a/src/freedreno/vulkan/tu_descriptor_set.cc b/src/freedreno/vulkan/tu_descriptor_set.cc
index bb6f5c653c2..ed9552f7a90 100644
--- a/src/freedreno/vulkan/tu_descriptor_set.cc
+++ b/src/freedreno/vulkan/tu_descriptor_set.cc
@@ -196,6 +196,17 @@ tu_CreateDescriptorSetLayout(
       set_layout->binding[b].offset = set_layout->size;
       set_layout->binding[b].dynamic_offset_offset = dynamic_offset_size;
       set_layout->binding[b].shader_stages = binding->stageFlags;
+      set_layout->binding[b].partially_bound =
+         /* Descriptor buffer implies PARTIALLY_BOUND. From a NOTE in the
+          * spec: "The requirements above imply that all descriptor bindings
+          * have been defined with the equivalent of ...
+          * VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT"
+          */
+         (pCreateInfo->flags &
+          VK_DESCRIPTOR_SET_LAYOUT_CREATE_DESCRIPTOR_BUFFER_BIT_EXT) ||
+         (variable_flags && j < variable_flags->bindingCount &&
+          (variable_flags->pBindingFlags[j] &
+           VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT));
 
       bool has_subsampled_sampler = false;
       if ((binding->descriptorType == VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER ||
@@ -473,6 +484,7 @@ blake3_update_descriptor_set_binding_layout(blake3_hasher *ctx,
    BLAKE3_UPDATE_VALUE(ctx, layout->array_size);
    BLAKE3_UPDATE_VALUE(ctx, layout->dynamic_offset_offset);
    BLAKE3_UPDATE_VALUE(ctx, layout->immutable_samplers_offset);
+   BLAKE3_UPDATE_VALUE(ctx, layout->partially_bound);
 
    const struct tu_sampler *samplers =
       tu_immutable_samplers(set_layout, layout);
diff --git a/src/freedreno/vulkan/tu_descriptor_set.h b/src/freedreno/vulkan/tu_descriptor_set.h
index 6a6a1ff9e59..1f6e52447a9 100644
--- a/src/freedreno/vulkan/tu_descriptor_set.h
+++ b/src/freedreno/vulkan/tu_descriptor_set.h
@@ -61,6 +61,12 @@ struct tu_descriptor_set_binding_layout
 
    /* Shader stages that use this binding */
    uint32_t shader_stages;
+
+   /* Whether statically accessing this binding guarantees that the descriptor
+    * accessed is valid. If true, only dynamically accessed descriptors are
+    * valid and we cannot speculate descriptor access.
+    */
+   bool partially_bound;
 };
 
 struct tu_descriptor_set_layout
-- 
2.55.0

